content/posts/draft-operator-constraints.markdown

```diff
@@ -0,0 +1,349 @@
+}
+
+func main() {
+  a := []int{42,23,1337}
+  less := IntComparator{}.Less // has type func(int, int) bool
+  slices.SortFunc(a, less)
+}
+```
+
+Based on this, we can devise a mechanism to allow custom comparisons:
+
+```go
+// Comparator is a helper type used to compare two T values.
+type Comparator[T any] interface {
+  ~struct{}
+  Less(a, b T) bool
+}
+
+func Sort[C Comparator[T], T any](a []T) {
+  var c C
+  less := c.Less // has type func(T, T) bool
+  // …
+}
+
+type SearchTree[C Comparator[T], T any] struct {
+  // …
+}
+```
+
+The `~struct{}` constraints any implementation of `Comparator[T]` to have
+underlying type `struct{}`. It is not strictly necessary, but it serves two
+purposes here:
+
+1. It makes clear that `Comparator[T]` itself is not supposed to carry any
+   state. It only exists to have its method called.
+2. It ensures (as much as possible) that the zero value of `C` is safe to use.
+   In particular, `Comparator[T]` would be a normal interface type. And it
+   would have a `Less` method of the right type, so it would implement itself.
+   But a zero `Comparator[T]` is `nil` and would always panic, if its method is
+   called.
+
+You can again provide helpers. This can also be used to combine this approach
+with the above ones:
+
+```go
+type LessOperator[T constraints.Ordered] struct{}
+
+func (LessOperator[T]) Less(a, b T) bool {
+  return a < b
+}
+
+type LessMethod[T Lesser[T]] struct{}
+
+func (LessMethod[T]) Less(a, b T) bool {
+  return a.Less(b)
+}
+
+type Reversed[C Comparator[T], T any] struct{}
+
+func (Reversed[C, T]) Less(a, b T) bool {
+  var c C
+  return c.Less(b, a)
+}
+```
+
+The advantage of this approach is that it makes the zero value of
+`SearchTree[C, T]` useful. For example, a `SearchTree[LessOperator[int], int]`
+can be used directly, without extra initialization.
+
+It also carries over the advantage of decoupling the comparison from the
+element type, which we got from accepting comparison functions.
+
+One disadvantage is that the comparator can never be inferred. It always has to
+be specified in the instantiation explicitly[^1]. That's similar to how we
+always had to pass a `less` function explicitly above.
+
+[^1]: Though as we put the `Comparator[T]` type parameter first, we can infer
+      `T` from the `Comparator`.
+
+Another disadvantage is that this *always* requires defining a type for
+comparisons. Where with the comparison function we could define customizations
+(like reversing the order) inline with a `func` literal, this mechanism always
+requires a method.
+
+Lastly, this is arguably too clever for its own good. Understanding the purpose
+and idea behind the `Comparator` type is likely to trip up your users when
+reading the documentation.
+
+## Summary
+
+We are left with these trade-offs:
+
+|| `constraints.Ordered` | `Lesser[T]` | `func(T,T) bool` | `Comparator[T]` |
+|--------------------------|----|--------|----|----|
+| Predeclared types        | 👍 | 👎     | 👎 | 👎 |
```

Comment:
The thumbs up and down are hard to differentiate on your page, maybe use ✅ and ❌?

Reply:
The checkmark and cross read as "yes or no" to me, while the thumbs read as "good or bad". That is, using the thumbs saves the extra step of having to decode which is "better". For example, for "Predeclared types" a "yes" is better, while for "type boilerplate" a "no" is better.

Alternatively, I would have to find phrasings for every line which mean "yes is always better", which also have to be short enough. I failed at that, myself :)